controllers/metal3.io/host_state_machine.go

```diff
@@ -227,6 +235,37 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
+func (hsm *hostStateMachine) checkDetachedHost(info *reconcileInfo) (result actionResult) {
+	// If the detached annotation is set we remove any host from the
+	// provisioner and take no further action
+	// Note this doesn't change the current state, only the OperationalStatus
+	annotations := hsm.Host.GetAnnotations()
+	if annotations != nil {
+		if _, ok := annotations[metal3v1alpha1.DetachedAnnotation]; ok {
+			// Only allow detaching hosts in Provisioned/ExternallyProvisioned states
+			if info.host.Status.Provisioning.State == metal3v1alpha1.StateProvisioned || info.host.Status.Provisioning.State == metal3v1alpha1.StateExternallyProvisioned {
+				return hsm.Reconciler.actionDetached(hsm.Provisioner, info)
+			}
+		}
+	}
+	if info.host.Status.ErrorType == metal3v1alpha1.DetachError {
+		clearError(info.host)
+		hsm.Host.Status.ErrorCount = 0
+		info.log.Info("removed detach error")
+		return actionUpdate{}
+	}
+	if info.host.OperationalStatus() == metal3v1alpha1.OperationalStatusDetached {
+		newStatus := metal3v1alpha1.OperationalStatusOK
+		if info.host.Status.ErrorType != "" {
```

Comment:
@zaneb with the change to explicitly handle DetachError as suggested by @andfasano I'm not sure this code is reachable?

In the case where we detach a host that's e.g `ProvisionedRegistrationError` either the Detach will succeed (so the ErrorType/Message gets removed), or fail (in which case this ends up DetachError)

I couldn't think of a way to reach this in the test without e.g forcing OperationalStatusDetached without any annotation, but if you can think of a scenario where we'll hit this in practice I can add another test scenario.

Reply:
It seems to me that if a detach succeeds (or is in the process of succeeding) then we should only clear the error if it is a DetachError.
For other operations we clear any error if any subsequent thing succeeds, on the principal that if something failed it was probably because we couldn't contact the hardware, so if it's back then time to start with a clean slate. But that doesn't really apply to Detach, which doesn't talk to the hardware (and also currently can't fail, although it's good that we're building all the integration so that it would work in theory).

Reply:
My point was that if we reach the point where `info.host.OperationalStatus() == metal3v1alpha1.OperationalStatusDetached` I'm not sure there will ever be any ErrorType except empty, because we clear errors via `clearErrorSetStatus` at the end of `actionDetached()`

Leaving this test for ErrorType is harmless, but I couldn't get test coverage without doing things which I don't think will ever actually happen, so I wanted to query if this logic is actually correct.

Reply:
Right, you're correct that as it stands I don't think we'll ever get here. What I'm suggesting is that we probably *shouldn't* clear any non-`DetachError` (which I think can only be a power management error, since we're only doing this in the steady states) at the end of (or during) the successful detach. If we only clear `DetachError`s in `actionDetached()` then this code will become reachable again.